File: android/src/main/java/com/facebook/flipper/plugins/inspector/descriptors/utils/stethocopies/FragmentCompat.java
/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook.flipper.plugins.inspector.descriptors.utils.stethocopies;

import android.app.Activity;
import android.content.Context;
import android.os.Build;
import android.util.Log;

import androidx.fragment.app.FragmentManager;

import java.lang.reflect.Field;
import java.util.Collections;
import java.util.List;
import javax.annotation.Nullable;
import javax.annotation.concurrent.NotThreadSafe;

/**
 * Compatibility abstraction which allows us to generalize access to both the support library's
 * fragments and the built-in framework version. Note: both versions can be live at the same time in
 * a single application and even on a single object instance.
 *
 * <p>Type safety is enforced via generics internal to the implementation but treated as opaque from
 * the outside.
 *
 * @param <FRAGMENT>
 * @param <DIALOG_FRAGMENT>
 * @param <FRAGMENT_MANAGER>
 * @param <FRAGMENT_ACTIVITY>
 */
@NotThreadSafe
public abstract class FragmentCompat<
    FRAGMENT, DIALOG_FRAGMENT, FRAGMENT_MANAGER, FRAGMENT_ACTIVITY extends Activity> {
  private static FragmentCompat sFrameworkInstance;
  private static FragmentCompat sSupportInstance;

  private static final boolean sHasSupportFragment;

  static {
    sHasSupportFragment =
        ReflectionUtil.tryGetClassForName("androidx.fragment.app.Fragment") != null;
  }

  @Nullable
  public static FragmentCompat getFrameworkInstance() {
    if (sFrameworkInstance == null && Build.VERSION.SDK_INT >= Build.VERSION_CODES.HONEYCOMB) {
      sFrameworkInstance = new FragmentCompatFramework();
    }
    return sFrameworkInstance;
  }

  @Nullable
  public static FragmentCompat getSupportLibInstance() {
    if (sSupportInstance == null && sHasSupportFragment) {
      sSupportInstance = new FragmentCompatSupportLib();
    }
    return sSupportInstance;
  }

  FragmentCompat() {}

  public abstract Class<FRAGMENT> getFragmentClass();

  public abstract Class<DIALOG_FRAGMENT> getDialogFragmentClass();

  public abstract Class<FRAGMENT_ACTIVITY> getFragmentActivityClass();

  public abstract FragmentAccessor<FRAGMENT, FRAGMENT_MANAGER> forFragment();

  public abstract DialogFragmentAccessor<DIALOG_FRAGMENT, FRAGMENT, FRAGMENT_MANAGER>
      forDialogFragment();

  public abstract FragmentManagerAccessor<FRAGMENT_MANAGER, FRAGMENT> forFragmentManager();

  public abstract FragmentActivityAccessor<FRAGMENT_ACTIVITY, FRAGMENT_MANAGER>
      forFragmentActivity();

  static class FragmentManagerAccessorViaReflection<FRAGMENT_MANAGER, FRAGMENT>
      implements FragmentManagerAccessor<FRAGMENT_MANAGER, FRAGMENT> {
    @Nullable private Field mFieldMAdded;

    @SuppressWarnings("unchecked")
    @Nullable
    @Override
    public List<FRAGMENT> getAddedFragments(FRAGMENT_MANAGER fragmentManager) {
      if (fragmentManager instanceof android.app.FragmentManager) {
        // This field is actually sitting on FragmentManagerImpl, which derives from
        // FragmentManager.
        if (mFieldMAdded == null) {
          Field fieldMAdded =
              ReflectionUtil.tryGetDeclaredField(fragmentManager.getClass(), "mAdded");

          if (fieldMAdded != null) {
            fieldMAdded.setAccessible(true);
            mFieldMAdded = fieldMAdded;
          }
        }
      } else if (fragmentManager instanceof androidx.fragment.app.FragmentManager) {
        return (List<FRAGMENT>) ((FragmentManager) fragmentManager).getFragments();
      }
      return Collections.emptyList();
    }
  }
}

File: android/src/main/java/com/facebook/flipper/plugins/uidebugger/stetho/FragmentCompat.kt
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook.flipper.plugins.uidebugger.stetho

import android.app.Activity
import android.view.View
import com.facebook.flipper.plugins.uidebugger.stetho.ReflectionUtil.getFieldValue
import com.facebook.flipper.plugins.uidebugger.stetho.ReflectionUtil.tryGetClassForName
import java.lang.reflect.Field
import javax.annotation.concurrent.NotThreadSafe

/**
 * Compatibility abstraction which allows us to generalize access to both the support library's
 * fragments and the built-in framework version. Note: both versions can be live at the same time in
 * a single application and even on a single object instance.
 *
 * Type safety is enforced via generics internal to the implementation but treated as opaque from
 * the outside.
 *
 * @param <FRAGMENT>
 * @param <DIALOG_FRAGMENT>
 * @param <FRAGMENT_MANAGER>
 * @param <FRAGMENT_ACTIVITY>
 */
@NotThreadSafe
abstract class FragmentCompat<
    FRAGMENT, DIALOG_FRAGMENT, FRAGMENT_MANAGER, FRAGMENT_ACTIVITY : Activity> {

  companion object {
    var frameworkInstance: FragmentCompat<*, *, *, *>? = null
      get() {
        if (field == null) {
          field = FragmentCompatFramework()
        }
        return field
      }

    var supportInstance: FragmentCompat<*, *, *, *>? = null
      get() {
        if (field == null && hasSupportFragment) {
          field = FragmentCompatSupportLib()
        }
        return field
      }

    private var hasSupportFragment = false
    init {
      hasSupportFragment = tryGetClassForName("androidx.fragment.app.Fragment") != null
    }

    fun isDialogFragment(fragment: Any): Boolean {
      val supportLib: FragmentCompat<*, *, *, *>? = supportInstance
      if (supportLib != null && supportLib.dialogFragmentClass?.isInstance(fragment) == true) {
        return true
      }

      val framework: FragmentCompat<*, *, *, *>? = frameworkInstance
      return framework != null && framework.dialogFragmentClass?.isInstance(fragment) == true
    }

    fun findFragmentForView(view: View): Any? {
      val activity = ViewUtil.tryGetActivity(view) ?: return null
      return findFragmentForViewInActivity(activity, view)
    }

    private fun findFragmentForViewInActivity(activity: Activity, view: View): Any? {
      val supportLib: FragmentCompat<*, *, *, *>? = supportInstance

      // Try the support library version if it is present and the activity is FragmentActivity.
      if (supportLib != null && supportLib.fragmentActivityClass?.isInstance(activity) == true) {
        val fragment = supportLib.findFragmentForViewInActivity(activity, view)
        if (fragment != null) {
          return fragment
        }
      }

      // Try the actual Android runtime version if we are on a sufficiently high API level for it to
      // exist.  Note that technically we can have both the support library and the framework
      // version in the same object instance due to FragmentActivity extending Activity (which has
      // fragment support in the system).
      val framework: FragmentCompat<*, *, *, *>? = frameworkInstance
      if (framework != null) {
        val fragment = framework.findFragmentForViewInActivity(activity, view)
        if (fragment != null) {
          return fragment
        }
      }
      return null
    }
  }

  abstract val fragmentClass: Class<FRAGMENT>?
  abstract val dialogFragmentClass: Class<DIALOG_FRAGMENT>?
  abstract val fragmentActivityClass: Class<FRAGMENT_ACTIVITY>?

  abstract fun forFragment(): FragmentAccessor<FRAGMENT, FRAGMENT_MANAGER>?
  abstract fun forDialogFragment():
      DialogFragmentAccessor<DIALOG_FRAGMENT, FRAGMENT, FRAGMENT_MANAGER>?
  abstract fun forFragmentManager(): FragmentManagerAccessor<FRAGMENT_MANAGER, FRAGMENT>?
  abstract fun forFragmentActivity(): FragmentActivityAccessor<FRAGMENT_ACTIVITY, FRAGMENT_MANAGER>?

  abstract fun getDialogFragments(activity: Activity): List<Any>
  abstract fun findFragmentForViewInActivity(activity: Activity, view: View): Any?
  abstract fun findFragmentForViewInFragment(fragment: Any, view: View): Any?

  class FragmentManagerAccessorViaReflection<FRAGMENT_MANAGER, FRAGMENT> :
      FragmentManagerAccessor<FRAGMENT_MANAGER, FRAGMENT> {
    private var fieldMAdded: Field? = null

    override fun getAddedFragments(fragmentManager: FRAGMENT_MANAGER): List<FRAGMENT> {

      // This field is actually sitting on FragmentManagerImpl, which derives from FragmentManager
      if (fieldMAdded == null) {
        fragmentManager?.let { manager ->
          val field = manager::class.java.getDeclaredField("mAdded")
          field.isAccessible = true
          fieldMAdded = field
        }
      }

      fieldMAdded?.let { field ->
        return getFieldValue(field, fragmentManager) as List<FRAGMENT>
      }

      return emptyList()
    }
  }
}
